paddle/framework/net_design.md

```diff
@@ -0,0 +1,159 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators in a network, 
+users can build a real network from a `NetDef` in protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetworkBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// The minimum a network should be implemented.
+class NetworkBase {
+ public:
+  // `def` is a proto message that describe the structure of a network.
+  NetworkBase(const NetDef& def);
+
+  // run all the operators and return success(true) or not, all the
+  // variables are located in `scope`.
+  virtual bool Run(Scope* scope) const = 0;
+
+ protected:
+  // keys of the input variables feed into the network.
+  std::vector<string> inputs_;
+  // keys of the corresponding output variables the network will mutate.
+  std::vector<string> outputs_;
+};
+```
+
+All network implementations should build networks from  a protobuf message which 
+describes the structure of a real network; `Run` method should be implemented by 
+all implementations to offer a universal method to forward or backward compute a network.
+
+A method of factory pattern can be defined like
+
+```c++
+std::unique<NetworkBase* CreateNet(const NetDef& def) {
+  switch (def.model_type()) {
+    case NN:
+      return new Network(def);
+    case Recursive:
+      return new RecursiveNet(def);
+    case Recurrent:
+      return new RecurrentNet(def);
+  }
+  return new Network(def);
+}
+```
+
+Network is designed as the container of operators, to make it more extendable,
```

Comment:
--> Network is designed as a container for operators, with related variable resources decoupled from it to make it more extendable.